src/models/mambafirstformer.py

```diff
@@ -0,0 +1,492 @@
+    attn_weights = torch.matmul(query, key.transpose(-1, -2))
+  
+    if self.scale_attn_weights:
+      attn_weights = attn_weights / (value.size(-1) ** 0.5)
+  
+      # Layer-wise attention scaling
+    if self.scale_attn_by_inverse_layer_idx:
+      attn_weights = attn_weights / float(self.layer_idx + 1)
+  
+    if not self.is_cross_attention:
+        # if only "normal" attention layer implements causal mask
+      query_length, key_length = query.size(-2), key.size(-2)
+      causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length].bool()
+      attn_weights = torch.where(causal_mask, attn_weights, self.masked_bias.to(attn_weights.dtype))
+  
+    if attention_mask is not None:
+        # Apply the attention mask
+        attn_weights = attn_weights + attention_mask
+  
+    attn_weights = nn.functional.relu(attn_weights) / query.size(-2) 
+  
+    # Downcast (if necessary) back to V's dtype (if in mixed-precision) -- No-Op otherwise
+    attn_weights = attn_weights.type(value.dtype)
+    attn_weights = self.attn_dropout(attn_weights)
+  
+    # Mask heads if we want to
+    if head_mask is not None:
+        attn_weights = attn_weights * head_mask
+  
+    attn_output = torch.matmul(attn_weights, value)
+  
+    return attn_output, attn_weights
+
+def relu_attn_causal(self, query, key, value, attention_mask=None, head_mask=None):
+    attn_weights = torch.matmul(query, key.transpose(-1, -2))
+  
+    if self.scale_attn_weights:
+        attn_weights = attn_weights / (value.size(-1) ** 0.5)
+  
+    # Layer-wise attention scaling
+    if self.scale_attn_by_inverse_layer_idx:
+        attn_weights = attn_weights / float(self.layer_idx + 1)
+  
+    if not self.is_cross_attention:
+        # if only "normal" attention layer implements causal mask
+        query_length, key_length = query.size(-2), key.size(-2)
+        causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length].bool()
+        attn_weights = torch.where(causal_mask, attn_weights, self.masked_bias.to(attn_weights.dtype))
+  
+    if attention_mask is not None:
+        # Apply the attention mask
+        attn_weights = attn_weights + attention_mask
+  
+      # TODO: make this sequence length causal (divide by tokens seen so far, not total tokens in sequence)
+      # relud = nn.functional.relu(attn_weights)
+    seq_len = query.size(-2)
+    causal_seq_len = 1 + ( torch.arange(seq_len, device=DEVICE)
+                                  .expand(attn_weights.shape)
+                                  .transpose(-1, -2) )
+      # import code
+      # assert attn_weights.shape == causal_seq_len.shape, code.interact(local=locals(), banner=f"Failed shape check: attn_weights do not math causal_seq_len in shape! \n{attn_weights.shape} vs {causal_seq_len.shape}")
+      # pre_attn_weights = attn_weights
+    attn_weights = nn.functional.relu(attn_weights) / (causal_seq_len + 1)
+      # code.interact(local=locals(), banner="yeesh")
```

Comment:
excess comments ... the TODO is out of sync -- that's what's happening here already